substreams/ethereum-balancer-v3/src/modules.rs

```diff
@@ -0,0 +1,503 @@
+        });
+
+    // Extract and insert any storage changes that happened for any of the components.
+    extract_contract_changes_builder(
+        &block,
+        |addr| {
+            components_store
+                .get_last(format!("pool:0x{0}", hex::encode(addr)))
+                .is_some() ||
+                addr.eq(VAULT_ADDRESS)
+        },
+        &mut transaction_changes,
+    );
+
+    transaction_changes
+        .iter_mut()
+        .for_each(|(_, change)| {
+            // this indirection is necessary due to borrowing rules.
+            let addresses = change
+                .changed_contracts()
+                .map(|e| e.to_vec())
+                .collect::<Vec<_>>();
+
+            addresses
+                .into_iter()
+                .for_each(|address| {
+                    if address != VAULT_ADDRESS {
+                        // We reconstruct the component_id from the address here
+                        let id = components_store
+                            .get_last(format!("pool:0x{}", hex::encode(address)))
+                            .map(|c| c.id)
+                            .unwrap(); // Shouldn't happen because we filter by known components
+                                       // in `extract_contract_changes_builder`
+                        change.mark_component_as_updated(&id);
+                    }
+                })
+        });
+
+    // Process all `transaction_changes` for final output in the `BlockChanges`,
+    //  sorted by transaction index (the key).
+    Ok(BlockChanges {
+        block: Some((&block).into()),
+        changes: transaction_changes
+            .drain()
+            .sorted_unstable_by_key(|(index, _)| *index)
+            .filter_map(|(_, builder)| builder.build())
+            .collect::<Vec<_>>(),
+    })
+}
+
+/// Converts address bytes into a Vec<u8> containing a leading `0x`.
+fn address_to_bytes_with_0x(address: &[u8; 20]) -> Vec<u8> {
+    address_to_string_with_0x(address).into_bytes()
+}
+
+/// Converts address bytes into a string containing a leading `0x`.
+fn address_to_string_with_0x(address: &[u8]) -> String {
+    format!("0x{}", hex::encode(address))
+}
+
+// function needed to match reservesOf in vault storage, which by definition
+// they should always be equal to the `token.balanceOf(this)` except during unlock
+fn get_vault_reserves(
+    transaction: &eth::v2::TransactionTrace,
+    store: &StoreGetProto<ProtocolComponent>,
+) -> HashMap<Vec<u8>, Vec<u8>> {
+    // reservesOf mapping for the current block Address -> Balance
+    let mut reserves_of = HashMap::<Vec<u8>, Vec<u8>>::new();
+    for call in &transaction.calls {
+        if let Some(Settle { token, .. }) = Settle::match_and_decode(call) {
+            for change in &call.storage_changes {
+                add_change_if_accounted(&mut reserves_of, change, token.as_slice());
+            }
+        }
+        if let Some(SendTo { token, .. }) = SendTo::match_and_decode(call) {
```

Comment:
What happens if in the same TX the user calls SendTo and then Settle? We'd keep the final value? Also, are you sure that these are the only OPs that can alter a vault's balance?

Reply:
SendTo requires the vault to be unlocked. Doesn't it also require `settle` to be called in the end?